File: sdk/csharp/tests/ConfigTests.cs
using Seocheon.Sdk.Errors;
using Xunit;

namespace Seocheon.Sdk.Tests;

public class ConfigTests
{
    private static SdkConfig ValidConfig() => new()
    {
        Chain = new ChainConfig
        {
            ChainId = "seocheon-1",
            RpcEndpoint = "http://localhost:26657",
            GrpcEndpoint = "http://localhost:1317"
        },
        Signing = new SigningConfig
        {
            Mode = SigningMode.Direct,
            Mnemonic = "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about"
        }
    };

    [Fact]
    public void ValidConfig_PassesValidation()
    {
        var config = ValidConfig();
        config.Validate(); // Should not throw
    }

    [Fact]
    public void MissingChainId_Throws()
    {
        var config = ValidConfig() with
        {
            Chain = ValidConfig().Chain with { ChainId = "" }
        };
        var ex = Assert.Throws<SdkException>(config.Validate);
        Assert.Equal(SdkErrors.CodeInvalidConfig, ex.Code);
    }

    [Fact]
    public void MissingRpcEndpoint_Throws()
    {
        var config = ValidConfig() with
        {
            Chain = ValidConfig().Chain with { RpcEndpoint = "" }
        };
        Assert.Throws<SdkException>(config.Validate);
    }

    [Fact]
    public void MissingGrpcEndpoint_Throws()
    {
        var config = ValidConfig() with
        {
            Chain = ValidConfig().Chain with { GrpcEndpoint = "" }
        };
        Assert.Throws<SdkException>(config.Validate);
    }

    [Fact]
    public void VaultMode_RequiresEndpointAndKeyName()
    {
        var config = ValidConfig() with
        {
            Signing = new SigningConfig { Mode = SigningMode.Vault }
        };
        Assert.Throws<SdkException>(config.Validate);
    }

    [Fact]
    public void KeystoreMode_RequiresPathAndPassphraseEnv()
    {
        var config = ValidConfig() with
        {
            Signing = new SigningConfig { Mode = SigningMode.Keystore }
        };
        Assert.Throws<SdkException>(config.Validate);
    }

    [Fact]
    public void DirectMode_RequiresMnemonic()
    {
        var config = ValidConfig() with
        {
            Signing = new SigningConfig { Mode = SigningMode.Direct }
        };
        Assert.Throws<SdkException>(config.Validate);
    }

    [Fact]
    public void InvalidBroadcastMode_Throws()
    {
        var config = ValidConfig() with
        {
            Tx = new TxConfig { BroadcastMode = "invalid" }
        };
        Assert.Throws<SdkException>(config.Validate);
    }
}

File: sdk/swift/Tests/SeocheonSDKTests/ConfigTests.swift
import XCTest
@testable import SeocheonSDK

final class ConfigTests: XCTestCase {

    func testValidDirectConfig() throws {
        let config = SDKConfig(
            chain: ChainConfig(chainId: "seocheon-1", rpcEndpoint: "http://localhost:26657", grpcEndpoint: "http://localhost:1317"),
            signing: SigningConfig(mode: .direct, mnemonic: "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about")
        )
        XCTAssertNoThrow(try config.validate())
    }

    func testValidVaultConfig() throws {
        let config = SDKConfig(
            chain: ChainConfig(chainId: "seocheon-1", rpcEndpoint: "http://localhost:26657", grpcEndpoint: "http://localhost:1317"),
            signing: SigningConfig(mode: .vault, vaultEndpoint: "http://vault:8200", keyName: "mykey")
        )
        XCTAssertNoThrow(try config.validate())
    }

    func testValidKeystoreConfig() throws {
        let config = SDKConfig(
            chain: ChainConfig(chainId: "seocheon-1", rpcEndpoint: "http://localhost:26657", grpcEndpoint: "http://localhost:1317"),
            signing: SigningConfig(mode: .keystore, keystorePath: "/path/to/keystore.json", passphraseEnv: "MY_PASS")
        )
        XCTAssertNoThrow(try config.validate())
    }

    func testMissingChainID() {
        let config = SDKConfig(
            chain: ChainConfig(chainId: "", rpcEndpoint: "http://localhost:26657", grpcEndpoint: "http://localhost:1317"),
            signing: SigningConfig(mode: .direct, mnemonic: "test mnemonic")
        )
        XCTAssertThrowsError(try config.validate()) { error in
            guard let sdkErr = error as? SDKError else { XCTFail("Expected SDKError"); return }
            XCTAssertEqual(sdkErr.code, 9005)
        }
    }

    func testMissingRPCEndpoint() {
        let config = SDKConfig(
            chain: ChainConfig(chainId: "seocheon-1", rpcEndpoint: "", grpcEndpoint: "http://localhost:1317"),
            signing: SigningConfig(mode: .direct, mnemonic: "test mnemonic")
        )
        XCTAssertThrowsError(try config.validate())
    }

    func testMissingMnemonic() {
        let config = SDKConfig(
            chain: ChainConfig(chainId: "seocheon-1", rpcEndpoint: "http://localhost:26657", grpcEndpoint: "http://localhost:1317"),
            signing: SigningConfig(mode: .direct)
        )
        XCTAssertThrowsError(try config.validate())
    }

    func testDefaultConfig() {
        let config = SDKConfig.defaultConfig(
            chainId: "seocheon-1",
            rpcEndpoint: "http://localhost:26657",
            grpcEndpoint: "http://localhost:1317",
            signing: SigningConfig(mode: .direct, mnemonic: "test mnemonic")
        )
        XCTAssertEqual(config.chain.chainId, "seocheon-1")
        XCTAssertEqual(config.tx.broadcastMode, "sync")
        XCTAssertEqual(config.tx.confirmTimeoutMs, 30000)
    }

    func testTxConfigDefaults() {
        let txConfig = TxConfig()
        XCTAssertEqual(txConfig.broadcastMode, "sync")
        XCTAssertEqual(txConfig.confirmTimeoutMs, 30000)
        XCTAssertEqual(txConfig.confirmPollIntervalMs, 1000)
    }
}
